New Examples Transform Command

StarKhan6368 opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Please make sure that you are familiar with and follow the Code of Conduct for
this project (found in the CODE_OF_CONDUCT.md file).

Also, please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->

**What**: Examples Transform Command

<!-- Why are these changes necessary? -->

**Why**: The examples transform command enables you to transform your examples based on the new OpenAPI specification. This specification is generated by applying the provided overlay to the original specification.

<!-- How were these changes implemented? -->

<!-- Have you done all of these things?  -->

**Checklist**:

<!-- add "N/A" to the end of each line that's irrelevant to your changes -->

<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->

- [ ] Documentation added to the README.md OR link to PR on https://github.com/specmatic/specmatic-documentation
- [x] Tests
- [ ] Sonar Quality Gate

<!-- feel free to add additional comments -->

yogeshnikam671 committed:
WIP

StarKhan6368 committed:
[WIP] Implement example transformation.

- New method in pattern to remove optional keys from value.

joelrosario commented on core/src/main/kotlin/io/specmatic/core/pattern/JSONObjectPattern.kt:
```diff
@@ -27,6 +27,19 @@ data class JSONObjectPattern(
+    override fun eliminateOptionalKey(value: Value, resolver: Resolver): Value {
+        if (value !is JSONObjectValue) return value
+
+        val filteredPattern = this.pattern.filterKeys { !it.endsWith("?") }
```
filteredPattern and filteredJsonObject variable names should be renamed for clarity. Perhaps mandatoryObjectPatternMap and mandatoryObjectMap.

StarKhan6368 replied:
Thanks, I've fixed that

joelrosario commented on core/src/main/kotlin/io/specmatic/core/pattern/JSONArrayPattern.kt:
```diff
@@ -13,6 +13,11 @@ import io.specmatic.core.value.Value
 data class JSONArrayPattern(override val pattern: List<Pattern> = emptyList(), override val typeAlias: String? = null) : Pattern, SequenceType {
+
+    override fun eliminateOptionalKey(value: Value, resolver: Resolver): Value {
+        return JSONArrayValue(emptyList())
```
Either we should fully implement, or not at all. Given this is a legacy pattern, I suggest not implementing for now.

StarKhan6368 replied:
Thanks, removed the implementation

joelrosario commented on core/src/main/kotlin/io/specmatic/core/value/JSONObjectValue.kt:
```diff
@@ -86,6 +86,21 @@ data class JSONObjectValue(val jsonObject: Map<String, Value> = emptyMap()) : Va
     fun findFirstChildByName(name: String): Value? =
         jsonObject[name]
+
+    fun objectWithMandatoryKeys(
```
Do we still need this method? I believe it's not being called from anywhere.

StarKhan6368 replied:
Agree, i've dropped the method now

StarKhan6368 committed:
implement feedback, code clean-ups.

- Remove eliminateOptionalKey for JSONArrayPattern.
- Better namings for variables in JSONObjectPattern.eliminateOptionalKey
- Remove unused method from JSONObjectValue.

joelrosario commented on application/src/main/kotlin/application/ExamplesCommand.kt:
```diff
@@ -534,6 +538,36 @@ For example:
+        lateinit var examplesDir: File
+
+        @Option(names = ["--only-mandatory-keys-in-payload"], description = ["Transform existing examples so that they contain only mandatory keys in payload"], required = false)
+        var allowOnlyMandatoryKeysInPayload: Boolean = false
```
this should be implicitly set to true if given, we should not have to say --only-mandatory-keys-in-payload=true

StarKhan6368 replied:
yeah makes sense, I've addressed this, also added a few simple tests for now.

joelrosario commented on core/src/main/kotlin/io/specmatic/core/examples/server/ExamplesInteractiveServer.kt:
```diff
@@ -724,6 +725,52 @@ class ExamplesInteractiveServer(
+            val examples = examplesDir.getExamplesFromDir()
+
+            examples.forEach { example ->
+                consoleLog("\nTransforming ${example.file.nameWithoutExtension}")
```
Put a verbose mode in, so we can switch the logging on if needed. There are too many logs if it's on by default.

StarKhan6368 committed:
default only-mandatory-keys to true, add tests.

- Simple tests in JSONObjectPattern, ListPattern and AnyPattern for
  eliminateOptionalKey method.

StarKhan6368 committed:
Fix - resolver not being updated with flags.

- Use flagBased updated resolver when eliminating Optional keys.
- Hide all transform example command logs behind a debug flag.
- Add consoleDebug helper functions.
